internal/graph/cached_resolver.go

```diff
@@ -170,32 +170,43 @@ func (c *CachedCheckResolver) ResolveCheck(
 }
 
 // checkRequestCacheKey converts the ResolveCheckRequest into a canonical cache key that can be
-// used for Check resolution cache key lookups.
-// The same tuple provided with the same contextual tuples should produce the same
-// cache key. If the contextual tuples are different order, it is possible that a different
-// cache key will be produced. This will result in duplicate entries.
+// used for Check resolution cache key lookups in a stable way.
+//
+// The same tuple provided with the same contextual tuples and context should produce the same
+// cache key. Contextual tuple order and context parameter order is ignored, only the contents
+// are compared.
 func checkRequestCacheKey(req *ResolveCheckRequest) (string, error) {
-	var contextualTuplesCacheKey string
+	hasher := keys.NewCacheKeyHasher(xxhash.New())
```

Comment:
curious to know more about what issues we had with the previous implementation and how this one solves it.

Reply:
Previous implementation had two major flaws:

1. `req.GetTupleKey().String()` is not guaranteed to produce a stable formatted string. If you follow the call chain, it utlimately calls `prototext.MarshalOptions{Multiline: false}.Format(m)`. According to the godoc for `.Format(..)`

> Format formats the message as a string.
> This method is only intended for human consumption and ignores errors.
> Do not depend on the output being stable. It may change over time across
> different versions of the program.

Depending on this for a cache key would not be recommended as it is not stable.

2. Contextual tuples would cache cache misses if the same contextual tuples were present but in different order. Taking a deterministic Hash64 of the context and contextual tuples produces a stable cache key regardless of order. Previously the following would lead to a cache miss.

```
Check(
  document:1#viewer@user:jon,
  contextual_tuples=["group:1#member@user:jon", "group:2#member@user:jon"]
) --> {allowed: true}, cache the result

Check(
  document:1#viewer@user:jon,
  contextual_tuples=["group:2#member@user:jon", "group:1#member@user:jon"]
) --> cache miss (because of contextual tuple order) --> {allowed: true}, cache the result under a different cache key, causing greater cache contention
```

`Check